File: python/nlolib/_requests.py
"""
Propagation request normalization for both low-level and high-level APIs.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable

from ._binding import (
    NLO_STORAGE_DB_CAP_POLICY_STOP_WRITES,
    NloExecutionOptions,
    NloPhysicsConfig,
    NloSimulationConfig,
)
from ._config import (
    PreparedSimConfig,
    _default_frequency_grid,
    _normalize_pulse_spec,
    _resolve_operator_spec,
    _solver_profile_defaults,
    _validate_coupled_pulse_spec,
    prepare_sim_config,
)
from ._models import ProgressInfo, RuntimeOperators


@dataclass
class _NormalizedPropagateRequest:
    sim_cfg: NloSimulationConfig
    phys_cfg: NloPhysicsConfig
    input_seq: list[complex]
    num_records: int
    exec_options: NloExecutionOptions | None
    sqlite_path: str | None
    run_id: str | None
    sqlite_max_bytes: int
    chunk_records: int
    cap_policy: int
    log_final_output_field_to_db: bool
    return_records: bool
    capture_step_history: bool
    step_history_capacity: int
    output_label: str
    explicit_record_z: list[float] | None
    progress_callback: Callable[[ProgressInfo], bool | int | None] | None
    meta_overrides: dict[str, Any]


class PropagateRequestBuilder:
    def from_config(
        self,
        config: PreparedSimConfig | NloSimulationConfig,
        *args: Any,
        **kwargs: Any,
    ) -> _NormalizedPropagateRequest:
        input_field = kwargs.pop("input_field", None)
        num_recorded_samples = kwargs.pop("num_recorded_samples", None)
        physics_config = kwargs.pop("physics_config", None)
        exec_options = kwargs.pop("exec_options", None)

        if len(args) > 3:
            raise TypeError("low-level propagate accepts at most three positional args after config")
        if len(args) >= 1:
            if input_field is not None:
                raise TypeError("input_field provided in both args and kwargs")
            input_field = args[0]
        if len(args) >= 2:
            if num_recorded_samples is not None:
                raise TypeError("num_recorded_samples provided in both args and kwargs")
            num_recorded_samples = args[1]
        if len(args) == 3:
            if exec_options is not None:
                raise TypeError("exec_options provided in both args and kwargs")
            exec_options = args[2]

        if input_field is None or num_recorded_samples is None:
            raise TypeError("low-level propagate requires input_field and num_recorded_samples")

        sqlite_path = kwargs.pop("sqlite_path", None)
        run_id = kwargs.pop("run_id", None)
        sqlite_max_bytes = int(kwargs.pop("sqlite_max_bytes", 0))
        chunk_records = int(kwargs.pop("chunk_records", 0))
        cap_policy = int(kwargs.pop("cap_policy", NLO_STORAGE_DB_CAP_POLICY_STOP_WRITES))
        log_final_output_field_to_db = bool(kwargs.pop("log_final_output_field_to_db", False))
        return_records = bool(kwargs.pop("return_records", True))
        capture_step_history = bool(kwargs.pop("capture_step_history", False))
        step_history_capacity = int(kwargs.pop("step_history_capacity", (200000 if capture_step_history else 0)))
        progress_callback = kwargs.pop("progress_callback", None)
        t_eval_raw = kwargs.pop("t_eval", None)
        if kwargs:
            raise TypeError(f"unexpected propagate kwargs: {sorted(kwargs.keys())}")

        num_records = int(num_recorded_samples)
        if num_records <= 0:
            raise ValueError("num_recorded_samples must be > 0")

        input_seq = list(input_field)
        if len(input_seq) == 0:
            raise ValueError("input_field must be non-empty")
        if step_history_capacity < 0:
            raise ValueError("step_history_capacity must be >= 0")
        if capture_step_history and step_history_capacity <= 0:
            raise ValueError("step_history_capacity must be > 0 when capture_step_history=True")

        explicit_record_z: list[float] | None = None
        if t_eval_raw is not None:
            explicit_record_z = [float(v) for v in t_eval_raw]
            if len(explicit_record_z) <= 0:
                raise ValueError("t_eval must be non-empty when provided")
            num_records = len(explicit_record_z)

        if isinstance(config, PreparedSimConfig):
            sim_cfg = config.simulation_config
            phys_cfg = config.physics_config
        else:
            sim_cfg = config
            phys_cfg = NloPhysicsConfig()
        if physics_config is not None:
            phys_cfg = physics_config

        return _NormalizedPropagateRequest(
            sim_cfg=sim_cfg,
            phys_cfg=phys_cfg,
            input_seq=input_seq,
            num_records=num_records,
            exec_options=exec_options,
            sqlite_path=(str(sqlite_path) if sqlite_path is not None else None),
            run_id=(None if run_id is None else str(run_id)),
            sqlite_max_bytes=sqlite_max_bytes,
            chunk_records=chunk_records,
            cap_policy=cap_policy,
            log_final_output_field_to_db=log_final_output_field_to_db,
            return_records=return_records,
            capture_step_history=capture_step_history,
            step_history_capacity=step_history_capacity,
            output_label=("final" if num_records == 1 else "dense"),
            explicit_record_z=explicit_record_z,
            progress_callback=progress_callback,
            meta_overrides={},
        )

    def from_pulse(
        self,
        pulse: Any,
        *args: Any,
        **kwargs: Any,
    ) -> _NormalizedPropagateRequest:
        if len(args) > 2:
            raise TypeError("high-level propagate accepts at most two positional operator args")

        linear_operator = kwargs.pop("linear_operator", "gvd")
        nonlinear_operator = kwargs.pop("nonlinear_operator", "kerr")
        if len(args) >= 1:
            linear_operator = args[0]
        if len(args) >= 2:
            nonlinear_operator = args[1]

        if "transverse_operator" in kwargs:
            raise TypeError(
                "transverse_operator has been removed; encode diffraction in linear_operator "
                "with tensor descriptors (tensor_nt/tensor_nx/tensor_ny)"
            )
        propagation_distance = kwargs.pop("propagation_distance", None)
        pulse_period_override = kwargs.pop("pulse_period", None)
        frequency_grid_override = kwargs.pop("frequency_grid", None)
        tensor_nt_override = kwargs.pop("tensor_nt", None)
        tensor_nx_override = kwargs.pop("tensor_nx", None)
        tensor_ny_override = kwargs.pop("tensor_ny", None)
        tensor_layout_override = kwargs.pop("tensor_layout", None)
        delta_x_override = kwargs.pop("delta_x", None)
        delta_y_override = kwargs.pop("delta_y", None)
        spatial_frequency_grid_override = kwargs.pop("spatial_frequency_grid", None)
        potential_grid_override = kwargs.pop("potential_grid", None)
        output = kwargs.pop("output", "dense")
        preset = kwargs.pop("preset", "balanced")
        records = kwargs.pop("records", None)
        starting_step_size_override = kwargs.pop("starting_step_size", None)
        max_step_size_override = kwargs.pop("max_step_size", None)
        min_step_size_override = kwargs.pop("min_step_size", None)
        error_tolerance_override = kwargs.pop("error_tolerance", None)
        exec_options = kwargs.pop("exec_options", None)
        sqlite_path = kwargs.pop("sqlite_path", None)
        run_id = kwargs.pop("run_id", None)
        sqlite_max_bytes = int(kwargs.pop("sqlite_max_bytes", 0))
        chunk_records = int(kwargs.pop("chunk_records", 0))
        cap_policy = int(kwargs.pop("cap_policy", NLO_STORAGE_DB_CAP_POLICY_STOP_WRITES))
        log_final_output_field_to_db = bool(kwargs.pop("log_final_output_field_to_db", False))
        return_records = bool(kwargs.pop("return_records", True))
        capture_step_history = bool(kwargs.pop("capture_step_history", False))
        step_history_capacity = int(kwargs.pop("step_history_capacity", (200000 if capture_step_history else 0)))
        progress_callback = kwargs.pop("progress_callback", None)
        t_eval_raw = kwargs.pop("t_eval", None)
        nonlinear_model_override = kwargs.pop("nonlinear_model", None)
        nonlinear_gamma_override = kwargs.pop("nonlinear_gamma", None)
        raman_fraction_override = kwargs.pop("raman_fraction", None)
        raman_tau1_override = kwargs.pop("raman_tau1", None)
        raman_tau2_override = kwargs.pop("raman_tau2", None)
        shock_omega0_override = kwargs.pop("shock_omega0", None)
        raman_response_time_override = kwargs.pop("raman_response_time", None)
        constants_override = kwargs.pop("constants", None)
        constant_bindings_override = kwargs.pop("constant_bindings", None)
        auto_capture_constants_override = kwargs.pop("auto_capture_constants", None)
        if kwargs:
            raise TypeError(f"unexpected high-level propagate kwargs: {sorted(kwargs.keys())}")
        if propagation_distance is None:
            raise TypeError("high-level propagate requires propagation_distance")
        if step_history_capacity < 0:
            raise ValueError("step_history_capacity must be >= 0")
        if capture_step_history and step_history_capacity <= 0:
            raise ValueError("step_history_capacity must be > 0 when capture_step_history=True")

        pulse_spec = _normalize_pulse_spec(pulse)
        if pulse_period_override is not None:
            pulse_spec.pulse_period = float(pulse_period_override)
        if frequency_grid_override is not None:
            pulse_spec.frequency_grid = frequency_grid_override  # type: ignore[assignment]
        if tensor_nt_override is not None:
            pulse_spec.tensor_nt = int(tensor_nt_override)
        if tensor_nx_override is not None:
            pulse_spec.tensor_nx = int(tensor_nx_override)
        if tensor_ny_override is not None:
            pulse_spec.tensor_ny = int(tensor_ny_override)
        if tensor_layout_override is not None:
            pulse_spec.tensor_layout = int(tensor_layout_override)
        if delta_x_override is not None:
            pulse_spec.delta_x = float(delta_x_override)
        if delta_y_override is not None:
            pulse_spec.delta_y = float(delta_y_override)
        if spatial_frequency_grid_override is not None:
            pulse_spec.spatial_frequency_grid = spatial_frequency_grid_override  # type: ignore[assignment]
        if potential_grid_override is not None:
            pulse_spec.potential_grid = potential_grid_override  # type: ignore[assignment]

        profile = _solver_profile_defaults(preset, float(propagation_distance))
        if starting_step_size_override is not None:
            profile["starting_step_size"] = float(starting_step_size_override)
        if max_step_size_override is not None:
            profile["max_step_size"] = float(max_step_size_override)
        if min_step_size_override is not None:
            profile["min_step_size"] = float(min_step_size_override)
        if error_tolerance_override is not None:
            profile["error_tolerance"] = float(error_tolerance_override)
        num_records = int(records) if records is not None else int(profile["records"])
        if output == "final":
            num_records = 1
        elif output != "dense":
            raise ValueError("output must be 'dense' or 'final'")
        if num_records <= 0:
            raise ValueError("records must be > 0")

        explicit_record_z: list[float] | None = None
        if t_eval_raw is not None:
            explicit_record_z = [float(v) for v in t_eval_raw]
            if len(explicit_record_z) <= 0:
                raise ValueError("t_eval must be non-empty when provided")
            num_records = len(explicit_record_z)

        linear_expr, linear_fn, linear_constants, linear_bindings = _resolve_operator_spec(
            "linear",
            linear_operator,
            0,
        )
        nonlinear_expr, nonlinear_fn, nonlinear_constants, nonlinear_bindings = _resolve_operator_spec(
            "nonlinear",
            nonlinear_operator,
            len(linear_constants),
        )

        binding_map: dict[str, float] = {}
        for source in (linear_bindings, nonlinear_bindings):
            if source is None:
                continue
            for key, value in source.items():
                existing = binding_map.get(key)
                if existing is not None and existing != value:
                    raise ValueError(f"conflicting callable param '{key}' across operators")
                binding_map[key] = value

        num_time_samples = len(pulse_spec.samples)
        tensor_mode = (
            pulse_spec.tensor_nt is not None and
            pulse_spec.tensor_nx is not None and
            pulse_spec.tensor_ny is not None
        )
        if tensor_mode:
            _validate_coupled_pulse_spec(pulse_spec)
        temporal_samples = (
            int(pulse_spec.tensor_nt)
            if pulse_spec.tensor_nt is not None and int(pulse_spec.tensor_nt) > 0
            else num_time_samples
        )
        pulse_period = (
            float(pulse_spec.pulse_period)
            if pulse_spec.pulse_period is not None
            else float(pulse_spec.delta_time) * float(temporal_samples)
        )
        frequency_grid = (
            list(pulse_spec.frequency_grid)
            if pulse_spec.frequency_grid is not None
            else _default_frequency_grid(temporal_samples, pulse_spec.delta_time)
        )

        runtime = RuntimeOperators(
            linear_factor_expr=linear_expr,
            nonlinear_expr=nonlinear_expr,
            linear_factor_fn=linear_fn,
            nonlinear_fn=nonlinear_fn,
            constants=[*linear_constants, *nonlinear_constants],
            constant_bindings=binding_map if binding_map else None,
            auto_capture_constants=(not binding_map),
        )
        if nonlinear_model_override is not None:
            runtime.nonlinear_model = int(nonlinear_model_override)
        if nonlinear_gamma_override is not None:
            runtime.nonlinear_gamma = float(nonlinear_gamma_override)
        if raman_fraction_override is not None:
            runtime.raman_fraction = float(raman_fraction_override)
        if raman_tau1_override is not None:
            runtime.raman_tau1 = float(raman_tau1_override)
        if raman_tau2_override is not None:
            runtime.raman_tau2 = float(raman_tau2_override)
        if shock_omega0_override is not None:
            runtime.shock_omega0 = float(shock_omega0_override)
        if raman_response_time_override is not None:
            runtime.raman_response_time = raman_response_time_override  # type: ignore[assignment]
        if constants_override is not None:
            runtime.constants = constants_override  # type: ignore[assignment]
        if constant_bindings_override is not None:
            runtime.constant_bindings = constant_bindings_override  # type: ignore[assignment]
        if auto_capture_constants_override is not None:
            runtime.auto_capture_constants = bool(auto_capture_constants_override)

        config = prepare_sim_config(
            num_time_samples,
            propagation_distance=float(propagation_distance),
            starting_step_size=float(profile["starting_step_size"]),
            max_step_size=float(profile["max_step_size"]),
            min_step_size=float(profile["min_step_size"]),
            error_tolerance=float(profile["error_tolerance"]),
            pulse_period=float(pulse_period),
            delta_time=float(pulse_spec.delta_time),
            tensor_nt=pulse_spec.tensor_nt,
            tensor_nx=pulse_spec.tensor_nx,
            tensor_ny=pulse_spec.tensor_ny,
            tensor_layout=int(pulse_spec.tensor_layout),
            frequency_grid=frequency_grid,
            delta_x=float(pulse_spec.delta_x),
            delta_y=float(pulse_spec.delta_y),
            spatial_frequency_grid=pulse_spec.spatial_frequency_grid,
            potential_grid=pulse_spec.potential_grid,
            runtime=runtime,
        )
        sim_cfg = config.simulation_config
        phys_cfg = config.physics_config
        return _NormalizedPropagateRequest(
            sim_cfg=sim_cfg,
            phys_cfg=phys_cfg,
            input_seq=list(pulse_spec.samples),
            num_records=num_records,
            exec_options=exec_options,
            sqlite_path=(str(sqlite_path) if sqlite_path is not None else None),
            run_id=(None if run_id is None else str(run_id)),
            sqlite_max_bytes=sqlite_max_bytes,
            chunk_records=chunk_records,
            cap_policy=cap_policy,
            log_final_output_field_to_db=log_final_output_field_to_db,
            return_records=return_records,
            capture_step_history=capture_step_history,
            step_history_capacity=step_history_capacity,
            output_label=output,
            explicit_record_z=explicit_record_z,
            progress_callback=progress_callback,
            meta_overrides={
                "preset": preset,
                "output": output,
                "coupled": bool(
                    tensor_mode and ((int(pulse_spec.tensor_nx) > 1) or (int(pulse_spec.tensor_ny) > 1))
                ),
            },
        )

File: julia/src/requests.jl
struct NormalizedPropagateRequest
    sim_cfg::SimulationConfig
    phys_cfg::PhysicsConfig
    input_seq::Vector{ComplexF64}
    keepalive::Vector{Any}
    num_records::Int
    exec_options::Union{Nothing, ExecutionOptions}
    sqlite_path::Union{Nothing, String}
    run_id::Union{Nothing, String}
    sqlite_max_bytes::Int
    chunk_records::Int
    cap_policy::Int
    log_final_output_field_to_db::Bool
    return_records::Bool
    capture_step_history::Bool
    step_history_capacity::Int
    output_label::String
    explicit_record_z::Union{Nothing, Vector{Float64}}
    progress_callback::Any
    meta_overrides::Dict{String, Any}
end

struct PropagateRequestBuilder
end

function _normalize_solver_aliases(kwargs::Base.Iterators.Pairs)
    out = Dict{Symbol, Any}(pairs(kwargs))
    if haskey(out, :t_span)
        t_span = out[:t_span]
        length(t_span) == 2 || throw(ArgumentError("t_span must have length 2"))
        Float64(t_span[1]) == 0.0 || throw(ArgumentError("t_span start must be 0.0"))
        out[:propagation_distance] = Float64(t_span[2]) - Float64(t_span[1])
        delete!(out, :t_span)
    end
    haskey(out, :first_step) && !haskey(out, :starting_step_size) && (out[:starting_step_size] = Float64(pop!(out, :first_step)))
    haskey(out, :max_step) && !haskey(out, :max_step_size) && (out[:max_step_size] = Float64(pop!(out, :max_step)))
    haskey(out, :min_step) && !haskey(out, :min_step_size) && (out[:min_step_size] = Float64(pop!(out, :min_step)))
    haskey(out, :rtol) && !haskey(out, :error_tolerance) && (out[:error_tolerance] = Float64(pop!(out, :rtol)))
    haskey(out, :output_mode) && !haskey(out, :output) && (out[:output] = pop!(out, :output_mode))
    return out
end

function from_config(::PropagateRequestBuilder, config, input_field;
                     num_recorded_samples,
                     physics_config = nothing,
                     exec_options = nothing,
                     sqlite_path = nothing,
                     run_id = nothing,
                     sqlite_max_bytes::Integer = 0,
                     chunk_records::Integer = 0,
                     cap_policy::Integer = NLO_STORAGE_DB_CAP_POLICY_STOP_WRITES,
                     log_final_output_field_to_db::Bool = false,
                     return_records::Bool = true,
                     capture_step_history::Bool = false,
                     step_history_capacity::Integer = (capture_step_history ? 200000 : 0),
                     progress_callback = nothing,
                     t_eval = nothing)
    num_records = Int(num_recorded_samples)
    num_records > 0 || throw(ArgumentError("num_recorded_samples must be > 0"))
    input_seq = ComplexF64.(collect(input_field))
    isempty(input_seq) && throw(ArgumentError("input_field must be non-empty"))
    keepalive = Any[]
    sim_cfg = config isa PreparedSimConfig ? config.simulation_config : config
    phys_cfg = config isa PreparedSimConfig ? config.physics_config : (physics_config === nothing ? PhysicsConfig() : physics_config)
    config isa PreparedSimConfig && append!(keepalive, config.keepalive)
    physics_config !== nothing && (phys_cfg = physics_config)
    explicit_record_z = t_eval === nothing ? nothing : _validate_explicit_record_z(t_eval, sim_cfg.propagation.propagation_distance)
    explicit_record_z !== nothing && (num_records = length(explicit_record_z))
    return NormalizedPropagateRequest(
        sim_cfg,
        phys_cfg,
        input_seq,
        keepalive,
        num_records,
        exec_options,
        isnothing(sqlite_path) ? nothing : String(sqlite_path),
        isnothing(run_id) ? nothing : String(run_id),
        Int(sqlite_max_bytes),
        Int(chunk_records),
        Int(cap_policy),
        Bool(log_final_output_field_to_db),
        Bool(return_records),
        Bool(capture_step_history),
        Int(step_history_capacity),
        num_records == 1 ? "final" : "dense",
        explicit_record_z,
        progress_callback,
        Dict{String, Any}(),
    )
end

function from_pulse(::PropagateRequestBuilder, pulse, linear_operator = "gvd", nonlinear_operator = "kerr"; kwargs...)
    options = _normalize_solver_aliases(kwargs)
    haskey(options, :propagation_distance) || throw(ArgumentError("high-level propagate requires propagation_distance"))
    pulse_spec = _normalize_pulse_spec(pulse)
    if haskey(options, :pulse_period)
        pulse_spec.pulse_period = Float64(pop!(options, :pulse_period))
    end
    haskey(options, :frequency_grid) && (pulse_spec.frequency_grid = _complex_vector_or_nothing(pop!(options, :frequency_grid)))
    haskey(options, :tensor_nt) && (pulse_spec.tensor_nt = Int(pop!(options, :tensor_nt)))
    haskey(options, :tensor_nx) && (pulse_spec.tensor_nx = Int(pop!(options, :tensor_nx)))
    haskey(options, :tensor_ny) && (pulse_spec.tensor_ny = Int(pop!(options, :tensor_ny)))
    haskey(options, :tensor_layout) && (pulse_spec.tensor_layout = Int(pop!(options, :tensor_layout)))
    haskey(options, :delta_x) && (pulse_spec.delta_x = Float64(pop!(options, :delta_x)))
    haskey(options, :delta_y) && (pulse_spec.delta_y = Float64(pop!(options, :delta_y)))
    haskey(options, :spatial_frequency_grid) && (pulse_spec.spatial_frequency_grid = _complex_vector_or_nothing(pop!(options, :spatial_frequency_grid)))
    haskey(options, :potential_grid) && (pulse_spec.potential_grid = _complex_vector_or_nothing(pop!(options, :potential_grid)))

    propagation_distance = Float64(pop!(options, :propagation_distance))
    preset = String(get(options, :preset, "balanced"))
    profile = _solver_profile_defaults(preset, propagation_distance)
    start_step = Float64(get(options, :starting_step_size, profile["starting_step_size"]))
    max_step = Float64(get(options, :max_step_size, profile["max_step_size"]))
    min_step = Float64(get(options, :min_step_size, profile["min_step_size"]))
    tol = Float64(get(options, :error_tolerance, profile["error_tolerance"]))
    output = String(get(options, :output, "dense"))
    num_records = output == "final" ? 1 : Int(get(options, :records, profile["records"]))
    output in ("dense", "final") || throw(ArgumentError("output must be 'dense' or 'final'"))
    num_records > 0 || throw(ArgumentError("records must be > 0"))

    tensor_mode = !isnothing(pulse_spec.tensor_nt) && !isnothing(pulse_spec.tensor_nx) && !isnothing(pulse_spec.tensor_ny)
    tensor_mode && _validate_coupled_pulse_spec(pulse_spec)
    temporal_samples = isnothing(pulse_spec.tensor_nt) ? length(pulse_spec.samples) : Int(pulse_spec.tensor_nt)
    pulse_period = isnothing(pulse_spec.pulse_period) ? pulse_spec.delta_time * temporal_samples : Float64(pulse_spec.pulse_period)
    frequency_grid = isnothing(pulse_spec.frequency_grid) ? _default_frequency_grid(temporal_samples, pulse_spec.delta_time) : pulse_spec.frequency_grid

    linear_expr, linear_constants = _resolve_operator_spec("linear", linear_operator, 0)
    nonlinear_expr, nonlinear_constants = _resolve_operator_spec("nonlinear", nonlinear_operator, length(linear_constants))
    runtime = RuntimeOperators(
        linear_factor_expr = linear_expr,
        nonlinear_expr = nonlinear_expr,
        nonlinear_model = Int(get(options, :nonlinear_model, NLO_NONLINEAR_MODEL_EXPR)),
        nonlinear_gamma = Float64(get(options, :nonlinear_gamma, 0.0)),
        raman_fraction = Float64(get(options, :raman_fraction, 0.0)),
        raman_tau1 = Float64(get(options, :raman_tau1, 0.0122)),
        raman_tau2 = Float64(get(options, :raman_tau2, 0.0320)),
        shock_omega0 = Float64(get(options, :shock_omega0, 0.0)),
        raman_response_time = get(options, :raman_response_time, nothing),
        constants = vcat(linear_constants, nonlinear_constants),
        constant_bindings = nothing,
        auto_capture_constants = false,
    )

    prepared = prepare_sim_config(length(pulse_spec.samples);
        propagation_distance = propagation_distance,
        starting_step_size = start_step,
        max_step_size = max_step,
        min_step_size = min_step,
        error_tolerance = tol,
        pulse_period = pulse_period,
        delta_time = pulse_spec.delta_time,
        tensor_nt = pulse_spec.tensor_nt,
        tensor_nx = pulse_spec.tensor_nx,
        tensor_ny = pulse_spec.tensor_ny,
        tensor_layout = pulse_spec.tensor_layout,
        frequency_grid = frequency_grid,
        delta_x = pulse_spec.delta_x,
        delta_y = pulse_spec.delta_y,
        spatial_frequency_grid = pulse_spec.spatial_frequency_grid,
        potential_grid = pulse_spec.potential_grid,
        runtime = runtime
    )

    explicit_record_z = haskey(options, :t_eval) ? _validate_explicit_record_z(options[:t_eval], propagation_distance) : nothing
    explicit_record_z !== nothing && (num_records = length(explicit_record_z))
    capture_step_history = Bool(get(options, :capture_step_history, false))
    step_history_capacity = Int(get(options, :step_history_capacity, capture_step_history ? 200000 : 0))

    return NormalizedPropagateRequest(
        prepared.simulation_config,
        prepared.physics_config,
        pulse_spec.samples,
        copy(prepared.keepalive),
        num_records,
        get(options, :exec_options, nothing),
        isnothing(get(options, :sqlite_path, nothing)) ? nothing : String(options[:sqlite_path]),
        isnothing(get(options, :run_id, nothing)) ? nothing : String(options[:run_id]),
        Int(get(options, :sqlite_max_bytes, 0)),
        Int(get(options, :chunk_records, 0)),
        Int(get(options, :cap_policy, NLO_STORAGE_DB_CAP_POLICY_STOP_WRITES)),
        Bool(get(options, :log_final_output_field_to_db, false)),
        Bool(get(options, :return_records, true)),
        capture_step_history,
        step_history_capacity,
        output,
        explicit_record_z,
        get(options, :progress_callback, nothing),
        Dict(
            "preset" => preset,
            "output" => output,
            "coupled" => Bool(tensor_mode && ((Int(pulse_spec.tensor_nx) > 1) || (Int(pulse_spec.tensor_ny) > 1))),
        ),
    )
end
